source/common/thread_local/thread_local_impl.cc

```diff
@@ -17,93 +17,121 @@ thread_local InstanceImpl::ThreadLocalData InstanceImpl::thread_local_data_;
 
 InstanceImpl::~InstanceImpl() {
   ASSERT(std::this_thread::get_id() == main_thread_id_);
-  ASSERT(shutdown_);
+  ASSERT(state_ == State::Shutdown);
   thread_local_data_.data_.clear();
 }
 
 SlotPtr InstanceImpl::allocateSlot() {
   ASSERT(std::this_thread::get_id() == main_thread_id_);
-  ASSERT(!shutdown_);
+  ASSERT(state_ != State::Shutdown);
 
   if (free_slot_indexes_.empty()) {
     SlotImplPtr slot(new SlotImpl(*this, slots_.size()));
-    auto wrapper = std::make_unique<Bookkeeper>(*this, std::move(slot));
-    slots_.push_back(wrapper->slot_.get());
-    return wrapper;
+    slots_.push_back(slot.get());
+    return slot;
   }
   const uint32_t idx = free_slot_indexes_.front();
   free_slot_indexes_.pop_front();
   ASSERT(idx < slots_.size());
   SlotImplPtr slot(new SlotImpl(*this, idx));
   slots_[idx] = slot.get();
-  return std::make_unique<Bookkeeper>(*this, std::move(slot));
+  return slot;
+}
+
+InstanceImpl::SlotImpl::SlotImpl(InstanceImpl& parent, uint64_t index)
+    : parent_(parent), index_(index), still_alive_guard_(std::make_shared<bool>()),
+      active_callback_guard_(
+          new bool,
```

Comment:
1. nit: `std::make_shared<bool>(false);`  though I am not sure if `false` is needed. 
2. `new bool()` or `new bool(false)`. Otherwise the value is not initialized.